### PR TITLE
feat: generalise descriptive stats table

### PR DESCRIPTION
**Pull request by MartinBernstorff**
- [X] I have considered whether this PR needs review, and requested a review if necessary.

# Notes for reviewers
@bokajgd I've had to abstract the descriptive stats table further to fit my use case. Having to retrain a model to get  `custom_columns` into the `EvalDataset` was simply too slow a feedback loop, that's entirely on me. 

I'm notifying you here to show my alternate implementation, but you don't need to review it :-)

@KennethEnevoldsen This is a general abstraction to create a descriptive stats table.

### Where to start?
To review this, I'd suggest pulling the branch and running `test_generate_descriptive_stats_table`It's an integration test/working example. The rest is just implementation and unit testing.

It's a complete rewrite of something we had implemented roughly already, hence the diff. Feel free to ignore the previous implementation.

### Where are all the doc strings?!
Looking back on e.g. `timeseriesflattener,` we have a bunch of redundant doc strings that don't add any information above the variable name. I've tried to add docstrings to the objects as relevant, let me know if you think we need more.

### Refactoring
I considered moving some of the logic into the spec objects, but couldn't make a nice implementation. Super open to suggestions, since it'll improve cohesion quite a bit.

### Terminology
Outputs a dataframe like this:
image

Where the terminology is:
image
Red: `Dataset` (e.g. "Train" and "Test")
Purple: `VariableGroup` (e.g. "Visits" and "Patients")
Green: `Variable` (two are highlighted in the above, showing that a variable can be coerced to categorical as well)

Signe/Sara will be nesting like Patient > Visit > Prediction time. E.g. each of their visits can have more than one prediction time. This lead to an even greater need for abstraction.

Let me know if you have any questions!

### CODE REVIEW

File: src/psycop_model_evaluation/utils.py [hunk 1]
```diff
@@ -8,12 +8,14 @@
...
 from datetime import date, datetime
 from pathlib import Path
-from typing import Any, Union
+from typing import Any, Optional, Union
 
 import dill as pkl
 import numpy as np
 import pandas as pd
 import wandb
+from pydantic import BaseModel as PydanticBaseModel
```
Comment (KennethEnevoldsen) [inline src/psycop_model_evaluation/utils.py:17]: Any reason for the rename?

**Review by KennethEnevoldsen — COMMENTED** [review]
One relevant comment regarding the pivot (I wouldn't want to maintain that as it is now) otherwise looks good.

File: src/psycop_model_evaluation/descriptive_stats_table.py [hunk 2]
```diff
@@ -1,342 +1,293 @@
...
-            Union[pd.DataFrame, wandb.Table]: Table 1.
-        """
-        if save_path is not None:
-            save_path.parent.mkdir(parents=True, exist_ok=True)
+        for col_df in dataset_row_vals[1:]:
+            row = row.merge(col_df, on="Title", how="outer")
 
-        if self.eval_dataset.age is not None:
-            age_stats = self._generate_age_stats()
+        rows.append(row)
```
[the remark this answers is not in the document]
Comment (MartinBernstorff) [reply src/psycop_model_evaluation/descriptive_stats_table.py:271]: Excellent, this is exactly the kind of comment I was looking for! Looking at it now.

File: src/psycop_model_evaluation/utils.py (hunk 1 above)
Comment (MartinBernstorff) [reply src/psycop_model_evaluation/utils.py:17]: We're redefining BaseModel in line 38. This is the suggested pattern for Pydantic - import their Basemodel under an alias, then redefine `BaseModel` with the config options you want :-)

**Pull request merged**